File: project-{topic}-optimized/src/middleware/AuthMiddleware.cc
#include "AuthMiddleware.h"
#include "utils/JwtManager.h"
#include <drogon/drogon.h>
#include <spdlog/spdlog.h>

void AuthMiddleware::doFilter(const drogon::HttpRequestPtr &req,
                              drogon::FilterCallback &&fcb,
                              drogon::FilterChainCallback &&fccb) {
    // Check for Authorization header
    std::string authHeader = req->getHeader("Authorization");
    if (authHeader.empty() || authHeader.rfind("Bearer ", 0) != 0) { // Check for "Bearer " prefix
        spdlog::warn("AuthMiddleware: No or invalid Authorization header found for path: {}", req->path());
        auto resp = drogon::HttpResponse::newHttpResponse();
        resp->setStatusCode(drogon::k401Unauthorized);
        resp->setBody("{\"code\":401,\"message\":\"Unauthorized: Missing or invalid token\"}");
        resp->setContentTypeCode(drogon::CT_APPLICATION_JSON);
        fcb(resp);
        return;
    }

    // Extract token
    std::string token = authHeader.substr(7); // "Bearer ".length() == 7

    // Verify token
    auto claims = JwtManager::getInstance().verifyToken(token);
    if (!claims) {
        spdlog::warn("AuthMiddleware: JWT verification failed for path: {}", req->path());
        auto resp = drogon::HttpResponse::newHttpResponse();
        resp->setStatusCode(drogon::k401Unauthorized);
        resp->setBody("{\"code\":401,\"message\":\"Unauthorized: Invalid or expired token\"}");
        resp->setContentTypeCode(drogon::CT_APPLICATION_JSON);
        fcb(resp);
        return;
    }

    // Token is valid, store user info in request context
    UserInfo userInfo;
    userInfo.isAuthenticated = true;
    userInfo.userId = std::stoll(claims.value()["userId"].asString());
    userInfo.username = claims.value()["username"].asString();
    userInfo.role = claims.value()["role"].asString();

    req->attributes()->insert(CURRENT_USER_INFO_KEY, userInfo);
    spdlog::info("AuthMiddleware: User {} (ID: {}) authenticated with role '{}' for path: {}",
                 userInfo.username, userInfo.userId, userInfo.role, req->path());

    // Continue to the next filter or controller
    fccb();
}

bool AuthMiddleware::hasRole(const drogon::HttpRequestPtr &req, const std::string& requiredRole) {
    if (req->attributes()->find(CURRENT_USER_INFO_KEY) == req->attributes()->end()) {
        spdlog::error("AuthMiddleware::hasRole: UserInfo not found in request attributes.");
        return false; // Should not happen if AuthMiddleware ran successfully
    }
    const auto& userInfo = req->attributes()->get<UserInfo>(CURRENT_USER_INFO_KEY);
    bool has = (userInfo.role == requiredRole);
    if (!has) {
        spdlog::warn("User {} (ID: {}) with role '{}' attempted to access resource requiring role '{}'.",
                     userInfo.username, userInfo.userId, userInfo.role, requiredRole);
    }
    return has;
}

File: project-{topic}-advanced/backend/src/middleware/auth.middleware.js
const jwt = require('jsonwebtoken');
const httpStatus = require('http-status');
const ApiError = require('../utils/ApiError');
const config = require('../config/config');
const catchAsync = require('../utils/catchAsync');
const { User } = require('../db/models');

const verifyToken = (token) => {
  return jwt.verify(token, config.jwt.secret);
};

const auth = (...requiredRoles) => catchAsync(async (req, res, next) => {
  let token;
  if (req.headers.authorization && req.headers.authorization.startsWith('Bearer')) {
    token = req.headers.authorization.split(' ')[1];
  }

  if (!token) {
    throw new ApiError(httpStatus.UNAUTHORIZED, 'Authentication token missing');
  }

  const payload = verifyToken(token);
  const user = await User.findByPk(payload.sub);

  if (!user) {
    throw new ApiError(httpStatus.UNAUTHORIZED, 'User not found or token invalid');
  }

  // Set user on request object
  req.user = user;

  // Check for required roles
  if (requiredRoles.length && !requiredRoles.includes(user.role)) {
    throw new ApiError(httpStatus.FORBIDDEN, 'Access forbidden: Insufficient permissions');
  }

  next();
});

module.exports = auth;